network/CMakeLists.txt

```diff
@@ -0,0 +1,74 @@
+project(network)
+
+set(PUBLIC_SOURCES
+  downloader/utils.cpp
+  downloader/utils.hpp
+  http/background_uploader.hpp
+  http/client.cpp
+  http/client.hpp
+  http/payload.hpp
+  http/request.cpp
+  http/request.hpp
+  http/thread.hpp
+  http/thread_callback.hpp
+  http/uploader.hpp
+  chunks_download_strategy.cpp
+  chunks_download_strategy.hpp
+  download_status.hpp
+  non_http_error_code.hpp
+  progress.hpp
+  servers_list.cpp
+  servers_list.hpp
+  socket.hpp
+)
+
+omim_add_library(${PROJECT_NAME} ${PUBLIC_SOURCES})
+
+target_sources(${PROJECT_NAME} PRIVATE
+  internal/http/file_request.hpp
```

Comment:
Dropping the internal folder may simplify the structure.

Reply:
An internal folder was added on purpose.
As I wrote in the description, Internal folder will hide all sources (and headers) that shouldn't be visible outside of the library.
This will help to organize our sources and keep the public interface clear.
In theory (it can be tested) it may also reduce the binary size or at least linking stage time

Reply:
With your proposed change, developers immediately see header files that are part of the public interface. However, it may be harder to start using something necessary that was hidden before in the internal folder.

In this specific case, IMO it is easier to keep "internal" files in their subfolders ("http"), the "internal" folder adds an unnecessary level to the sources tree.

Reply:
I can propose another file structure:
```
├── include
│   ├── public_header_1.hpp
│   └── public_header_2.hpp
├── src
│   ├── private_header.hpp
│   └── source.cpp
├── tests
│   └── tests sources
└── CMakeLists.txt
   ```

But anyway, for a library there should be 
1. public interface
2. private implementation

> However, it may be harder to start using something necessary that was hidden before in the internal folder.

The only reason it may be harder (regarding our project) is that the feature is not yet implemented for every platform. We have this problem right now. Some files/classes are used in the iOS project, but they are not implemented for Android. So, there is no single public interface for it.
This should be refactored/fixed but not in this PR.
There are also some files that are not used at all (and I doubt they will be used in the future) but I didn't want to remove them. At least in this PR.

All in all, I think the problem you described is not a real problem :)
1. It's not hard to create a public header for a new feature. Because it's just a header, right?
2. You won't use something "necessary" that was implemented before (and wasn't already used) and lacks public headers because:
  1. If it lacks a public header then it's not fully implemented
  2. You cannot use something that is not implemented, you need to implement it first
  3. When you implement it, there is no problem with adding a public header for it.

P.S. I just reviewed these two file structures and must admit I like the second one more (with separate `include` and `src` folders).

@biodranik @vng WDYT?